modules/dag-shared/src/main/scala/org/tessellation/dag/validators/TransactionValidator.scala

```diff
@@ -0,0 +1,208 @@
+package org.tessellation.dag.validators
+
+import cats.data.ValidatedNel
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.traverse._
+import cats.syntax.validated._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.{Transaction, TransactionReference}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.key.ops.PublicKeyOps
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.signature.signature.SignatureProof
+
+import eu.timepit.refined.auto._
+import io.estatico.newtype.ops._
+
+trait TransactionValidator[F[_]] {
+  import TransactionValidator._
+
+  implicit val F: Async[F]
+  implicit val kryoSerializer: KryoSerializer[F]
+  implicit val securityProvider: SecurityProvider[F]
+
+  protected def getBalance(address: Address): F[Balance]
+
+  protected def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def validateSingleTransaction(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Signed[Transaction]]] =
+    for {
+      propertiesValidation <- validateTransactionProperties(signedTransaction)
+      balanceValidation <- validateSourceBalances(Seq(signedTransaction.value))
+    } yield propertiesValidation.product(balanceValidation).map(_ => signedTransaction)
+
+  def validateBatchOfTransactions(
+    signedTransactions: Seq[Signed[Transaction]]
+  ): F[ValidationResult[Seq[Signed[Transaction]]]] =
+    for {
+      propertiesValidation <- signedTransactions.traverse(validateTransactionProperties).map(_.sequence)
+      balancesValidation <- validateSourceBalances(signedTransactions.map(_.value))
+    } yield propertiesValidation.product(balancesValidation).map(_ => signedTransactions)
```

Comment:
for public API maybe following would be better?

```scala
def validate(transaction: Signed[Transaction]) = validate(Seq(transaction))
def validate(transactions: Seq[Signed[Transaction]]) = implementation
```

Then, you just use it like:
```
TransactionValidator.validate(tx)
TransactionValidator.validate(Seq(tx1, tx2))
```

Imo there is no point to use single/batch and "transaction" in the method name in this case

Reply:
Sure, makes sense.